substrate/primitives/runtime-interface/src/pass_by.rs

```diff
@@ -15,419 +15,698 @@
+	}
+}
+
+/// Pass a value into the host by a thin pointer.
 ///
-/// # Example
+/// This casts the value into a `&[u8]` using `AsRef<[u8]>` and passes a pointer to that byte blob
+/// to the host. Then the host reads `N` bytes from that address into an `[u8; N]`, converts it
+/// into target type using `From<[u8; N]>` and passes it into the host function by a reference.
 ///
-/// ```
-/// # use sp_runtime_interface::pass_by::PassByEnum;
-/// #[derive(PassByEnum, Copy, Clone)]
-/// enum Data {
-///     Okay,
-///     NotOkay,
-///     // This will not work with the derive.
-///     //Why(u32),
-/// }
-/// ```
-pub use sp_runtime_interface_proc_macro::PassByEnum;
-
-/// Something that should be passed between wasm and the host using the given strategy.
+/// This can only be used with reference types (e.g. `&[u8; 32]`). Use [`PassPointerAndReadCopy`]
+/// if you want to have the host function accept a non-reference type on the host side.
 ///
-/// See [`Codec`], [`Inner`] or [`Enum`] for more information about the provided strategies.
-pub trait PassBy: Sized {
-	/// The strategy that should be used to pass the type.
-	type PassBy: PassByImpl<Self>;
+/// Raw FFI type: `u32` (a pointer)
+pub struct PassPointerAndRead<T, const N: usize>(PhantomData<(T, [u8; N])>);
+
+impl<'a, T, const N: usize> RIType for PassPointerAndRead<&'a T, N> {
+	type FFIType = u32;
+	type Inner = &'a T;
 }
 
-/// Something that provides a strategy for passing a type between wasm and the host.
-///
-/// This trait exposes the same functionality as [`crate::host::IntoFFIValue`] and
-/// [`crate::host::FromFFIValue`] to delegate the implementation for a type to a different type.
-///
-/// This trait is used for the host implementation.
-#[cfg(feature = "std")]
-pub trait PassByImpl<T>: RIType {
-	/// Convert the given instance to the ffi value.
-	///
-	/// For more information see: [`crate::host::IntoFFIValue::into_ffi_value`]
-	fn into_ffi_value(instance: T, context: &mut dyn FunctionContext) -> Result<Self::FFIType>;
+#[cfg(not(substrate_runtime))]
+impl<'a, T, const N: usize> FromFFIValue<'a> for PassPointerAndRead<&'a T, N>
+where
+	T: From<[u8; N]>,
+{
+	type Owned = T;
+
+	fn from_ffi_value(
+		context: &mut dyn FunctionContext,
+		arg: Self::FFIType,
+	) -> Result<Self::Owned> {
+		let mut out = [0; N];
+		context.read_memory_into(Pointer::new(arg), &mut out)?;
+		Ok(T::from(out))
+	}
+
+	#[inline]
+	fn take_from_owned(owned: &'a mut Self::Owned) -> Self::Inner {
+		&*owned
+	}
+}
+
+#[cfg(substrate_runtime)]
+impl<'a, T, const N: usize> IntoFFIValue for PassPointerAndRead<&'a T, N>
+where
+	T: AsRef<[u8]>,
+{
+	type Destructor = ();
 
-	/// Create `T` from the given ffi value.
-	///
-	/// For more information see: [`crate::host::FromFFIValue::from_ffi_value`]
-	fn from_ffi_value(context: &mut dyn FunctionContext, arg: Self::FFIType) -> Result<T>;
+	fn into_ffi_value(value: &mut Self::Inner) -> (Self::FFIType, Self::Destructor) {
+		assert_eq!(value.as_ref().len(), N);
```

Comment:
As a note, I think `Vec<u8>` satisfies both bound, maybe someone could write `PassPointerAndRead<&Vec<u8>, 4>`. Then the assertion that value is of length 4 would fail.